### PR TITLE
New ruler endpoints

### PR DESCRIPTION
**Pull request by jml**
Part of #619 

- creates new endpoints for getting & setting ruler configs; puts these endpoints on the ruler itself
  - these endpoints only exposed if db flags provided
- creates new database methods that use the existing data structure but mask the fact that ruler & alertmanager configs are mushed together
- updates ruler to use database for getting rules if provided right flag; falls back to config server otherwise

I was going to do alertmanager in the same PR, but this got big quickly.

I'm not wedded to putting the endpoints on the ruler itself.

Pros:
- we can delete the configs service
- the ruler scheduler doesn't need to poll the database

Cons:
- if the ruler scheduler slows down, setting & getting configs on the UI will be slower (although given they both talk to the same DB, their performance will be correlated anyhow)

I say "ruler scheduler", because I'm intending to split the scheduler from workers as soon as #619 is done.

One thing I'll insist on is that these endpoints with new behaviours be at different URLs to the ones currently being served.

### CODE REVIEW

**Comment by jml** [comment]
Just remembered I want to move over some API tests and run them against in memory and Postgres databases. The query over jsonb data is quite funky and I could easily get it wrong.

File: pkg/ruler/api.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		return
+	}
+	logger := util.WithContext(r.Context(), util.Logger)
+
+	cfg, err := a.db.GetRulesConfig(userID)
+	if err == sql.ErrNoRows {
+		http.Error(w, "No configuration", http.StatusNotFound)
+		return
+	} else if err != nil {
+		// XXX: Untested
```
Comment (bboreham) [inline pkg/ruler/api.go]: What is the expectation with this comment?

Comment (jml) [reply pkg/ruler/api.go]: It's cargo-culted from the config API. The vague expectation is that we'd test the error path one of these days. I've removed it & its like.

**Comment by jml** [comment]
Still needs some more tests before I'm happy to merge it.

**Comment by jml** [comment]
Also changed new API to CAS, which addresses part of #330

File: pkg/ruler/configs.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+// configsClient allows retrieving recording and alerting rules from the configs server.
+type configsClient struct {
+	URL     *url.URL
+	Timeout time.Duration
+}
+
+// GetConfigs implements RulesAPI.
+func (c configsClient) GetConfigs(since configs.ID) (map[string]configs.VersionedRulesConfig, error) {
+	suffix := ""
+	if since != 0 {
```
[the remark this answers is not in the document]
Comment (jml) [reply pkg/ruler/configs.go]: It's for rules, not alertmanager configs, but "why doesn't this use `RulerConfigsAPI.GetConfigs()`?" is a valid question. The answer is that it replaces `RulerConfigsAPI`, and has a slightly different API (`configs.VersionedRulesConfig` is new in this PR).

We're replacing it because we don't want to have a client library per se. The ruler will be directly responsible for querying the database, so there's no need for code in another package that knows how to get ruler configs.

**Comment by jml** [comment]
Wrote the tests. I'm personally happy for this to be merged. 

@bboreham PTAL?

File: pkg/ruler/api_test.go [hunk 3]
```diff
@@ -43,22 +44,22 @@ func cleanup(t *testing.T) {
 }
 
 // request makes a request to the configs API.
-func request(t *testing.T, method, urlStr string, body io.Reader) *httptest.ResponseRecorder {
+func request(t *testing.T, handler http.Handler, method, urlStr string, body io.Reader) *httptest.ResponseRecorder {
```
Comment (bboreham) [inline pkg/ruler/api_test.go]: Please include this change in the commit description, or in a separate commit if that makes sense.

File: pkg/configs/db/db.go [hunk 4]
```diff
@@ -24,12 +24,12 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 
 // RulesDB has ruler-specific DB interfaces.
 type RulesDB interface {
-	// GetRulesConfig gets the user's alertmanager config
+	// GetRulesConfig gets the user's ruler config
```
Comment (bboreham) [inline pkg/configs/db/db.go]: Please include this change in the commit description, or move to a separate commit if that makes more sense.

File: pkg/configs/db/postgres/postgres.go [hunk 5]
```diff
@@ -144,16 +145,27 @@ func (d DB) GetRulesConfig(userID string) (configs.VersionedRulesConfig, error)
...
+	updated := false
+	err := d.Transaction(func(tx DB) error {
+		current, err := d.GetConfig(userID)
+		if err != nil && err != sql.ErrNoRows {
+			return err
+		}
+		// The supplied oldConfig must match the current config. If no config
+		// exists, then oldConfig must be nil. Otherwise, it must exactly
+		// equal the existing config.
+		if !((err == sql.ErrNoRows && oldConfig == nil) || reflect.DeepEqual(current.Config.RulesFiles, oldConfig)) {
```
Comment (bboreham) [inline pkg/configs/db/postgres/postgres.go]: Less repetition if you moved the details of the check into a `RulesConfig.Equal()` method, and that could be coded as a straightforward loop rather than using `reflect`.

File: pkg/configs/db/postgres/postgres.go [hunk 6]
```diff
@@ -134,6 +135,86 @@ func (d DB) GetConfigs(since configs.ID) (map[string]configs.View, error) {
...
+	})
+	return updated, err
+}
+
+func (d DB) findRulesConfigs(filter squirrel.Sqlizer) (map[string]configs.VersionedRulesConfig, error) {
+	rows, err := d.Select("id", "owner_id", "config ->> 'rules_files'").
+		Options("DISTINCT ON (owner_id)").
+		From("configs").
+		Where(filter).
+		Where("config ->> 'rules_files' <> '{}'").
```
Comment (bboreham) [inline pkg/configs/db/postgres/postgres.go]: Why?

File: pkg/configs/db/postgres/postgres.go [hunk 7]
```diff
@@ -134,6 +135,86 @@ func (d DB) GetConfigs(since configs.ID) (map[string]configs.View, error) {
...
+		OrderBy("owner_id, id DESC").
```
Comment (bboreham) [inline pkg/configs/db/postgres/postgres.go]: Why don't we have a `Limit(1)` here like in `GetConfig()` ?

File: pkg/configs/db/db.go [hunk 8]
```diff
@@ -22,8 +22,24 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	flag.StringVar(&cfg.MigrationsDir, "database.migrations", "", "Path where the database migration files can be found")
 }
 
+// RulesDB has ruler-specific DB interfaces.
+type RulesDB interface {
+	// GetRulesConfig gets the user's ruler config
+	GetRulesConfig(userID string) (configs.VersionedRulesConfig, error)
+	// SetRulesConfig sets the user's ruler config
```
Comment (bboreham) [inline pkg/configs/db/db.go]: Mention the CAS-ness of the call

File: pkg/configs/db/postgres/postgres.go (hunk 5 above)
Comment (jml) [reply pkg/configs/db/postgres/postgres.go]: FWIW, in my early morning fog, it wasn't that straightforward. I have become too used to just writing `deriving Eq`.

File: pkg/configs/db/postgres/postgres.go (hunk 6 above)
Comment (jml) [reply pkg/configs/db/postgres/postgres.go]: Good question. I've added a fairly detailed comment, repasted here in case that assists your follow-up review.

```go
		// `->>` gets a JSON object field as text. When a config row exists
		// and alertmanager config is provided but ruler config has not yet
		// been, the 'rules_files' key will have an empty JSON object as its
		// value. This is (probably) the most efficient way to test for a
		// non-empty `rules_files` key.
		//
		// This whole situation is way too complicated. See
		// https://github.com/weaveworks/cortex/issues/619 for the whole
		// story, and our plans to improve it.
```

File: pkg/configs/db/postgres/postgres.go (hunk 7 above)
Comment (jml) [reply pkg/configs/db/postgres/postgres.go]: Because `GetConfig()` gets the configuration for a single user, whereas `findRulesConfigs` helps `GetAllRulesConfigs()` and `GetRulesConfigs()` get the active configurations for all users. 

I've added a doc comment that highlights the purpose, but doesn't explicitly answer the "why no `Limit(1)`?" question:

> // findRulesConfigs helps GetAllRulesConfigs and GetRulesConfigs retrieve the
> // set of all active rules configurations across all our users.

I'm open to renaming some or all of these functions if you can suggest names that more clearly communicate intent.

File: pkg/ruler/api_test.go (hunk 3 above)
Comment (jml) [reply pkg/ruler/api_test.go]: I've split this into a different commit.

File: pkg/configs/db/db.go (hunk 4 above)
Comment (jml) [reply pkg/configs/db/db.go]: I've moved it to be part of 029f17c, which introduces this interface.

**Comment by jml** [comment]
@bboreham I've addressed all your review points. I *think* I could sensibly split this into two PRs, where the first PR goes from master..3dc22b9 (`Public RulesAPI no longer used. Delete it.`) and changes the ruler to go directly to the database for getting all configs, and the second PR has the rest and adds the new endpoints & tests. I only want to do this if it accelerates getting both merged (as I think this is blocking #629).

PTAL.

File: pkg/configs/db/postgres/postgres.go (hunk 6 above)
Comment (bboreham) [reply pkg/configs/db/postgres/postgres.go]: This seems to assume we only ever go from blank to non-blank in the history; if it goes from non-blank to blank then you will select the previous version.

File: pkg/configs/db/postgres/postgres.go [hunk 9]
```diff
@@ -134,6 +134,97 @@ func (d DB) GetConfigs(since configs.ID) (map[string]configs.View, error) {
...
+		updated = true
+		return d.SetConfig(userID, new)
+	})
+	return updated, err
+}
+
+// findRulesConfigs helps GetAllRulesConfigs and GetRulesConfigs retrieve the
+// set of all active rules configurations across all our users.
+func (d DB) findRulesConfigs(filter squirrel.Sqlizer) (map[string]configs.VersionedRulesConfig, error) {
+	rows, err := d.Select("id", "owner_id", "config ->> 'rules_files'").
```
Comment (bboreham) [inline pkg/configs/db/postgres/postgres.go]: OK, now I don't understand why this doesn't bring back the entire history, so you will end up with some random version.

Comment (bboreham) [reply pkg/configs/db/postgres/postgres.go]: ah, "`SELECT DISTINCT ON ( expression [, ...] )` keeps only the first row of each set of rows"
and you have order by `id DESC` below.

So effectively the same as `LIMIT 1` elsewhere

Comment (jml) [reply pkg/configs/db/postgres/postgres.go]: Want me to add a comment along those lines? If so, can you share the link to the doc you just quoted?

Comment (bboreham) [reply pkg/configs/db/postgres/postgres.go]: https://www.postgresql.org/docs/9.5/static/sql-select.html#SQL-DISTINCT
